cmd/simulateSam/simulateSam.go

```diff
@@ -0,0 +1,89 @@
+// Command Group: "Data Simulation"
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"github.com/vertgenlab/gonomics/exception"
+	"github.com/vertgenlab/gonomics/fasta"
+	"github.com/vertgenlab/gonomics/fileio"
+	"github.com/vertgenlab/gonomics/sam"
+	"github.com/vertgenlab/gonomics/simulate"
+	"log"
+	"math/rand"
+)
+
+func usage() {
+	fmt.Print(
+		"simulateSam - Simulate alignments to a reference sequence\n\n" +
+			"Currently only generates illumina-style paired sequencing data" +
+			"Read Length : 150bp, Average Insert Size: 50bp" +
+			"Usage:\n" +
+			" simulateSam [options] -i ref.fasta\n" +
+			"options:\n")
+	flag.PrintDefaults()
+}
+
+func simulateSam(refFile, outFile string, numReads int, bamOutput bool) {
+	ref := fasta.Read(refFile)
+	out := fileio.EasyCreate(outFile)
+	header := sam.GenerateHeader(fasta.ToChromInfo(ref), nil, sam.Unsorted, sam.None)
+
+	var bw *sam.BamWriter
+	if bamOutput {
+		bw = sam.NewBamWriter(out, header)
+	} else {
+		sam.WriteHeaderToFileHandle(out, header)
+	}
+
+	var totalLen int
+	for i := range ref {
+		totalLen += len(ref[i].Seq)
+	}
+
+	var proportionOfReads float64
+	var readsPerContig int
+	var reads []sam.Sam
+	for i := range ref {
+		proportionOfReads = float64(len(ref[i].Seq)) / float64(totalLen)
+		readsPerContig = int(float64(numReads) * proportionOfReads)
+		reads = simulate.IlluminaSam(ref[i].Name, ref[i].Seq, readsPerContig)
```

Comment:
I'm not sure I follow the relationship between coverage and fragment length. Currently reads are distributed between chroms based on the length of the contig. A contig of size 2 gets twice as many reads as a contig of size 1. How does the fragment length factor in, and could/should it be added in the main function instead of the IlluminaSam function?

Reply:
I think there are two things here that are related, but mixed together.
1) How fragment length factors in.  I was thinking that you need to use fragment length to know the probability of getting a read from a section of DNA of a certain length.  For example, I would have thought that a 100bp chromosome would have a 0 percent chance of giving you a fragment that was 300bp long, but I now realize that in this case you add random bases to the end to allow a chromosome of any size to give a fragment of any size.  I should have gone back to my comments after realizing this and modified it.
2) Exact same number of reads on each chromosome.  This will likely be variable in real life and by fixing the simulation to always give the most likely number of reads (and not vary) the simulation may not have as much variance in read coverage as one might see in real world data.  For example.  I think chromosome 22 is about 1 to 2% of the human genome.  If I am understanding the code correctly, if I were to repeatedly simulate 50 reads from the human genome, I may never get a read from chromosome 22.  This is probably made to simulate very large numbers of reads, so this is not a huge concern, but there may be some corner case where the behavior is a bit odd by fixing the simulation to always give the expected number.

Reply:
Ah I understand now. The most recent commit distributes reads randomly, but weighted by contig length.